freshmaker/handlers/botas/botas_shipped_advisory.py

```diff
@@ -96,63 +133,144 @@ def handle(self, event):
+
+        if not bundle_digests_by_related_nvr:
+            msg = "None of the original images have related bundles, skip."
+            log.warning(msg)
+            db_event.transition(EventState.SKIPPED, msg)
+            return []
+
+        # A set of unique bundle digests, an image can be in multiple bundles' related_images
         bundle_digests = set()
-        for bundle in bundles:
-            if not bundle.get('bundle_path_digest'):
-                log.warning("Bundle %s doesn't have 'bundle_path_digests' set",
-                            bundle['bundle_path'])
+        for digests in bundle_digests_by_related_nvr.values():
+            bundle_digests.update(digests)
+
+        # Mapping of bundle digest to bundle data
+        # {
+        #     digest: {
+        #         "images": [image_amd64, image_aarch64],
+        #         "nvr": NVR,
+        #         "auto_rebuild": True/False,
+        #         "osbs_pinning": True/False,
+        #         "pullspecs": [...],
+        #     }
+        # }
+        bundles_by_digest = {}
+        default_bundle_data = {
+            'images': [],
+            'nvr': None,
+            'auto_rebuild': False,
+            'osbs_pinning': False,
+            'pullspecs': [],
+        }
+
+        # Get images for each bundle digest, a bundle digest can has multiple images
+        # with different arches.
+        for digest in bundle_digests:
+            bundles = self._pyxis.get_images_by_digest(digest)
+            # If no bundle image found, just skip this bundle digest
+            if not bundles:
+                continue
+
+            bundles_by_digest.setdefault(digest, copy.deepcopy(default_bundle_data))
+            bundles_by_digest[digest]['nvr'] = bundles[0]['brew']['build']
+            bundles_by_digest[digest]['images'] = bundles
+
+        # Unauthenticated koji session to fetch build info of bundles
+        koji_api = KojiService(conf.koji_profile)
+
+        # For each bundle, check whether it should be rebuilt by comparing the
+        # auto_rebuild_tags of repository and bundle's tags
+        for digest, bundle_data in bundles_by_digest.items():
+            bundle_nvr = bundle_data['nvr']
+
+            # Images are for different arches, just check against the first image
+            image = bundle_data['images'][0]
+            if self.image_has_auto_rebuild_tag(image):
+                bundle_data['auto_rebuild'] = True
+
+            # Fetch buildinfo
+            buildinfo = koji_api.get_build(bundle_nvr)
+            related_images = (
+                buildinfo.get("extra", {})
+                .get("image", {})
+                .get("operator_manifests", {})
+                .get("related_images", {})
+            )
+            bundle_data['osbs_pinning'] = related_images.get('created_by_osbs', False)
+            # Save the original pullspecs
+            bundle_data['pullspecs'] = related_images.get('pullspecs', [])
```

Comment:
I don't think it can happen, every published bundle image should has pullspecs of related images, and even with that can happen, we'll skip this bundle in later for loop:
```
    for pullspec in bundle_data['pullspecs']:
    ...
```